### ISSUE TITLE
Truncating error messages is hard for small errors

### ISSUE THREAD

**Issue by joneshf** [dhall-lang/dhall-haskell#861]
I know that seeing a giant error message is painful, but truncating the output is painful in a different way for small error messages.

For example, dealing with a record with three fields:
```Dhall
{ bin : List { file : Text, name : Text }, name : Text, sourceDirectory : Text }
```

If I give Dhall the wrong fields and things are small enough, I can see what the problem is:

```
You tried to invoke the following function:                                     
                                                                                
↳ < Haskell :
      { executables :
          List { executableDirectory : Text, executableName : Text }
  ==================================================================
          Text
      }
  >.JavaScript
                                                                                
... which expects an argument of type or kind:                                  
                                                                                
↳ { bin : List { file : Text, name : Text }, name : Text, sourceDirectory : Text }
                                                                                
... on the following argument:                                                  
                                                                                
↳ { name = "build-browserify", sourceDirectory = "src", version = "1.0.0" }
                                                                                
... which has a different type or kind:                                         
                                                                                
↳ { name : Text, sourceDirectory : Text, version : Text }
```

If I add the missing field `bin` (but don't remove the extra field `version`), and `bin` is longer than one line, I can't easily see what the problem is:

```
You tried to invoke the following function:                                     
                                                                                
↳ < Haskell :
      { executables :
          List { executableDirectory : Text, executableName : Text }
  ==================================================================
          Text
      }
  >.JavaScript
                                                                                
... which expects an argument of type or kind:                                  
                                                                                
↳ { bin : List { file : Text, name : Text }, name : Text, sourceDirectory : Text }
                                                                                
... on the following argument:                                                  
                                                                                
↳ { bin =
      [ { file = "src/index.js", name = "build-browserify" } ]
  , name =
  ============================================================
  , version =
      "1.0.0"
  }
                                                                                
... which has a different type or kind:                                         
                                                                                
↳ { bin :
      List { file : Text, name : Text }
  , name :
  =====================================
  , version :
      Text
  }

```

It works out in this case because I put `version` as the extra field. If `path` is the extra field, I get a really cryptic error:
```
You tried to invoke the following function:                                     
                                                                                
↳ < Haskell :
      { executables :
          List { executableDirectory : Text, executableName : Text }
  ==================================================================
          Text
      }
  >.JavaScript
                                                                                
... which expects an argument of type or kind:                                  
                                                                                
↳ { bin : List { file : Text, name : Text }, name : Text, sourceDirectory : Text }
                                                                                
... on the following argument:                                                  
                                                                                
↳ { bin =
      [ { file = "src/index.js", name = "build-browserify" } ]
  , name =
  ============================================================
  , sourceDirectory =
      "src"
  }
                                                                                
... which has a different type or kind:                                         
                                                                                
↳ { bin :
      List { file : Text, name : Text }
  , name :
  =====================================
  , sourceDirectory :
      Text
  }
```

All of the context I need to figure out the problem has been truncated away.

Showing everything is painful when the context is huge. Choosing an arbitrary number of lines to truncate out is impossible when the context is slightly over that number. Is there some compromise we can come to? I know we're focused on making large config files easy to deal with, but we're going to turn people away before they gain momentum if a record of three fields gives this sort of output.

_Originally posted by @joneshf in https://github.com/dhall-lang/dhall-haskell/issues/573#issuecomment-474467637_

**Comment by joneshf** [comment]
Also copying my follow-up reply: 

> Maybe the solution here isn't to focus on truncating? The real issue is that the record has the wrong fields. Maybe we can provide a better error message specific for records?
> 
> In the last example, if you don't use `--explain`, you get this output:
> 
> ```
> Error: Wrong type of function argument
> 
> { + path : …
> , …
> }
> 
>       Package.JavaScript
>       { bin =
>           [ { file = "src/index.js", name = "build-browserify" } ]
>       ============================================================
>           "src"
>       }
>     
> ```
> That
> 
> ```
> { + path : …
> , …
> }
> ```
> message is a bit hard to parse. Does it mean that I need to add `path`? If so, what's its type? Does it mean I need to remove `path`? It's especially hard when it's a nested record of who knows how deep:
> 
> ```
> Error: Wrong type of function argument
> 
> { + path : …
> ,   bin : …
>           { name : - Text
>                    + { … : … }
>           , …
>           }
> , …
> }
> 
>       Package.JavaScript
>       { bin =
>           [ { file =
>       ==================
>           "src"
>       }
> ```
> What's the issue here? How should I go about solving it? Should I throw some empty record in there and see what happens? Is there being a record the problem?
> 
> The current error message gives you clues about the problem, but it doesn't give you solutions to the problem. If we could say (in prose) that field `foo` is not in the record, so add it with the type `Foo`, that would be a step up. If we could say that `bar` is the right field, but it's got the wrong type and this is what the type should be, that would be a step up.
> 
> I recognize that we can't always know what the solution is. The error message could still be a heuristic for fixing the problem. If we say, "it's likely that doing X will solve the problem," that seems better than saying, "here are the things that are wrong, good luck!"
> 
> I also recognize that `--explain` provides a ton of information in the general case about the problem. That's super useful if you don't understand the general issue. Once you do understand general issue, you need a bit more help to solve the specific problem.

**Comment by jneira** [comment]
Quoting myself:

> I think truncation should be replaced with the "path" to the offending piece of code, using the lines of code of each nested definition (if present). I am afraid that, in practice, i've not found so useful the truncated version anyway.

But the full one is simply intractable, only wait to the output to end is overwhelming. So we need another way to focus in the problematic type/value, navigating into the nested data structures.

Not sure if haskell is a good example to follow but it tries:

```
D:\dev\lang\eta\eta\compiler\Eta\Main\HscMain.hs:1316:44:
    Couldn't match expected type ‘ModLocation’ with actual type ‘[a1]’
    In the third argument of ‘codeGen’, namely ‘location’
    In a stmt of a 'do' block:
      modClasses <- codeGen
                      hsc_env
                      this_mod
                      location
                      data_tycons
                      stg_binds
                      (panic "hpc_info")
                      (lookupStubs modClass foreign_stubs)
```

**Comment by Gabriel439** [comment]
One change I've made in the interim is that we now show much more of the expression before truncating (up to 40 lines).  See: https://github.com/dhall-lang/dhall-haskell/pull/916

**Comment by Gabriel439** [comment]
I'll close this because I believe it was addressed by #916.  Feel free to reopen if there is still something that needs to be fixed
